src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -168,6 +266,259 @@ private static IDataTransform Create(IHostEnvironment env, Arguments args, IData
+                    int rows = 0;
+                    bool isDataLeft = false;
+                    using (var ch = _host.Start("Training TensorFlow model..."))
+                    {
+                        while (cursor.MoveNext())
+                        {
+                            rows++;
+                            for (int i = 0; i < inputColIndices.Length; i++)
+                            {
+                                isDataLeft = true;
+                                srcTensorGetters[i].BufferTrainingData();
+                            }
+
+                            if ((rows % args.BatchSize) == 0)
+                            {
+                                isDataLeft = false;
+                                (loss, metric) = TrainBatch(inputColIndices, inputsForTraining, srcTensorGetters, fetchList, args);
+                            }
+                        }
+                        if (isDataLeft)
+                        {
+                            isDataLeft = false;
+                            ch.Warning("Not training on the last batch. The batch size is less than {0}.", args.BatchSize);
+                        }
+                        ch.Info("Loss: {0}, Metric: {1}", fetchList.Count > 0 ? loss.ToString("#.####") : "None",
+                                fetchList.Count > 1 ? metric.ToString("#.####") : "None");
+                    }
+                }
+            }
+            UpdateModelOnDisk(model, args);
+        }
+
+        private (float loss, float metric) TrainBatch(int[] inputColIndices,
+            string[] inputsForTraining,
+            ITensorValueGetter[] srcTensorGetters,
+            List<string> fetchList,
+            TrainingArguments args)
+        {
+            float loss = 0;
+            float metric = 0;
+            var runner = Session.GetRunner();
+            for (int i = 0; i < inputColIndices.Length; i++)
+            {
+                var inputName = inputsForTraining[i];
+                runner.AddInput(inputName, srcTensorGetters[i].GetBufferedBatchTensor());
+            }
+            if (args.LearningRateOperation != null)
+                runner.AddInput(args.LearningRateOperation, new TFTensor(args.LearningRate));
+            runner.AddTarget(args.OptimizationOperation);
+
+            if (fetchList.Count > 0)
+                runner.Fetch(fetchList.ToArray());
+
+            var tensor = runner.Run();
+            loss = tensor.Length > 0 ? (float)tensor[0].GetValue() : 0.0f;
+            metric = tensor.Length > 1 ? (float)tensor[1].GetValue() : 0.0f;
+
+            return (loss, metric);
+        }
+
+        /// <summary>
+        /// Updates the model on the disk.
+        /// After retraining Session and Graphs are both up-to-date
+        /// However model on disk is not which is used to serialzed to ML.Net stream
+        /// </summary>
+        private void UpdateModelOnDisk(string modelDir, TrainingArguments args)
```

Comment:
>UpdateModelOnDisk [](start = 21, length = 17)

all operations on FileSystem is extremely error prune, and can gave you exception for multiple reasons. And we prefer to mark exception in our code, so If it's possible I would prefer to have big try catch(ex) throw.Exception(ex); on this method.